In [1]:
import pandas as pd
import math
import numpy as np

## Factorial

In [3]:
def factorial(x):
    if x == 1:
        return 1
    else:
        return x*factorial(x-1)

In [5]:
factorial(3)

6

In [6]:
factorial(0)

RecursionError: maximum recursion depth exceeded in comparison

In [7]:
math.factorial(0)

1

In [11]:
def true_factorial(x):
    
    def factorial(x):
        if x == 1:
            return 1
        else:
            return x*factorial(x-1)
    
    return int(factorial(x+1)/(x+1))
    

In [15]:
true_factorial(5)

120

n! is the number of ways we can arrange n objects

$$ x! = x \times (x-1) \times (x-2) \times ... \times 2 \times 1 $$ 

$$ x! = \frac{(x+1)!}{(x+1)} = \frac{(x+1) \times x \times (x-1) \times (x-2) \times ... \times 2 \times 1}{x+1} $$ \
$$ 4! = \frac{5!}{5} = \frac{120}{5} = 24 $$ \
$$ 3! = \frac{4!}{4} = \frac{24}{4} = 6 $$ \
$$ 2! = \frac{3!}{3}  = \frac{6}{3} = 2  $$ \
$$ 1! = \frac{2!}{2} = \frac{2}{2} = 1 $$ \
$$ 0! = \frac{1!}{1} = \frac{1}{1} = 1 $$

## Fibonacci numbers

The Fibonacci numbers, commonly denoted F(n) form a sequence, called the Fibonacci sequence, such that each number is the sum of the two preceding ones, starting from 0 and 1.

$$ F(0) = 0,   \ $$
$$ F(1) = 1 $$ \
$$ F(N) = F(N - 1) + F(N - 2), for N > 1 $$


In [25]:
def fibonacci(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)

In [26]:
list(fibonacci(i) for i in range(10))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [2]:
def fib(n):
    if n in range(2):
        return n
    else:
        return fib(n-1) + fib(n-2)

In [16]:
for i in range(10):
    print(fib(i))

0
1
1
2
3
5
8
13
21
34


In [19]:
list(fib(i) for i in range(10))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [5]:
fib(2)

1

In [22]:
def pisano_period(d, length):
    fib_sequence = list(fib(i) for i in range(length))
    remainders = [num%d for num in fib_sequence]
    
    return remainders

In [23]:
pisano_period(5, 20)

[0, 1, 1, 2, 3, 0, 3, 3, 1, 4, 0, 4, 4, 3, 2, 0, 2, 2, 4, 1]

## Towers of Hanoi

Towers of Hanoi is a mathematical game with the following rules. Given three rods and n disks of different diameters, we need to move the entire stack of disks from one rod to another. However, there are three rules:
1. A bigger disk can never be on top of a smaller one. For example, you cannot put a disk with d=3 on top of a disk with d=1.
2. You can only move one disk at a time, so you can't just grab the entire stack and put it on a different rod.
3. You can't grab a disk from underneath another one. You can only move the one that is on top.

In [3]:
def towers_of_hanoi(n, source_rod, destination_rod, auxilliary_rod):
    if n == 1:
        print("Move disk 1 from rod", source_rod ,"to rod", destination_rod)

In [4]:
towers_of_hanoi(1, 'A', 'B', 'C')

Move disk 1 from rod A to rod B


In [9]:
def towers_of_hanoi(n, source_rod, destination_rod, auxilliary_rod):
    if n == 1:
        print("Move disk 1 from rod", source_rod,"to rod", destination_rod)
        return
    towers_of_hanoi(n-1, source_rod, auxilliary_rod, destination_rod) 
    print("Move disk", n, "from ", source_rod, "to ", destination_rod)
    towers_of_hanoi(n-1, auxilliary_rod, destination_rod, source_rod) 

In [10]:
towers_of_hanoi(3, 'A', 'B', 'C')

Move disk 1 from rod A to rod B
Move disk 2 from  A to  C
Move disk 1 from rod B to rod C
Move disk 3 from  A to  B
Move disk 1 from rod C to rod A
Move disk 2 from  C to  B
Move disk 1 from rod A to rod B


In [13]:
len('Eleven')

6